Code/GraphMol/ChemReactions/Enumerate/EvenSamplePairs.cpp

```diff
@@ -0,0 +1,278 @@
+      if (j == i) continue;
+      size_t ii = digits[i] + ioffset;
+      size_t jj = digits[j] + joffset;
+      if (pair_used[ii][jj] > 0) {
+        double numer = (double)pair_used[ii][jj];
+        double denom = sqrt((double)(rgroups[i]) * (double)(rgroups[j]));
+        islack = (int)(numer / denom);
+      }
+      joffset += rgroups[j];
+    }
+    ioffset += rgroups[i];
+  }
+
+  if (islack > nslack) {
+    rejected_bb_sampling_condition += 1;
+    return false;
+  }
+
+  // keep track of bb usage
+  for (size_t i = 0; i < num_rgroups; ++i) {
+    if (var_used[i][digits[i]] == 0) {
+      used_count[i]++;
+    }
+    var_used[i][digits[i]] += 1;
+    if (used_count[i] == rdcast<int64_t>(rgroups[i])) {
+      // complete variable scan => initialize
+      if (nslack > min_nslack && rgroups[i] > 1)  // cleared slack on i
+        nslack = min_nslack;
+
+      used_count[i] = 0;
+      for (size_t j = 0; j < rgroups[i]; ++j) {
+        var_used[i][j]--;
+        assert(var_used[i][j] >= 0);
+        if (var_used[i][j] > 0) used_count[i]++;
+      }
+    }  // end scan
+  }
+
+  // keep track of BB Pair usage
+  ioffset = 0;
+  for (size_t i = 0; i < num_rgroups; ioffset += rgroups[i], ++i) {
+    size_t joffset = 0;
+    for (size_t j = 0; j < num_rgroups; joffset += rgroups[j], ++j) {
+      if (j == i) {
+        continue;
+      }
+      size_t ii = digits[i] + ioffset;
+      size_t jj = digits[j] + joffset;
+      if (pair_used[ii][jj] == 0) {
+        pair_counts[i][j]++;
+      }
+      pair_used[ii][jj]++;
+      if (pair_counts[i][j] >= rgroups[i] * rgroups[j]) {  // all pairs visited
+        if (nslack > min_nslack && (rgroups[i] > 1 || rgroups[j] > 1)) {
+          nslack = min_nslack;
+        }
+        pair_counts[i][j] = 0;
+        for (size_t ii = 0; ii < rgroups[i]; ++ii) {
+          for (size_t jj = 0; jj < rgroups[j]; ++jj) {
+            pair_used[ioffset + ii][joffset + jj]--;
+            if (pair_used[ioffset + ii][joffset + jj] > 0) {
+              pair_counts[i][j]++;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  selected.insert(seed);
+  return true;
+}
+
+const RGROUPS &EvenSamplePairsStrategy::next() {
+  nslack = 0;
+  while (m_numPermutationsProcessed < rdcast<size_t>(m_numPermutations)) {
+    bool added = false;
+    for (size_t l = 0; l < M; ++l) {
+      seed = ((seed * a + b) % M);
+      if (seed > rdcast<size_t>(m_numPermutations)) {
+        rejected_period += 1;
+        continue;
+      } else if (selected.find(seed) != selected.end()) {
+        rejected_unique += 1;
+        continue;
+      } else if (try_add(seed)) {
+        m_numPermutationsProcessed++;
+        added = true;
+        return decode(seed);
+      }
+    }
+
+    if (!added) {
+      // loosen heuristic
+      nslack += 1;
+      min_nslack += 1;
+    }
+  }
+
+  throw EnumerationStrategyException("Ran out of molecules");
+}
+
+void EvenSamplePairsStrategy::stats() const {
```

Comment:
not necessary, but maybe useful: have this take the stream it should output to as an argument?